js/ai/src/generateAction.ts

```diff
@@ -0,0 +1,299 @@
+                new GenerateResponseChunk(chunk, accumulatedChunks)
+              );
+            }
+          }
+        : undefined,
+      async () => new GenerateResponse(await model(request))
+    );
+
+    // throw NoValidCandidates if all candidates are blocked or
+    if (
+      !response.candidates.some((c) =>
+        ['stop', 'length'].includes(c.finishReason)
+      )
+    ) {
+      throw new NoValidCandidatesError({
+        message: `All candidates returned finishReason issues: ${JSON.stringify(response.candidates.map((c) => c.finishReason))}`,
+        response,
+      });
+    }
+
+    if (input.output?.jsonSchema && !response.toolRequests()?.length) {
+      // find a candidate with valid output schema
+      const candidateErrors = response.candidates.map((c) => {
+        // don't validate messages that have no text or data
+        if (c.text() === '' && c.data() === null) return null;
+
+        try {
+          parseSchema(c.output(), {
+            jsonSchema: input.output?.jsonSchema,
+          });
+          return null;
+        } catch (e) {
+          return e as Error;
+        }
+      });
+      // if all candidates have a non-null error...
+      if (candidateErrors.every((c) => !!c)) {
+        throw new NoValidCandidatesError({
+          message: `Generation resulted in no candidates matching provided output schema.${candidateErrors.map((e, i) => `\n\nCandidate[${i}] ${e!.toString()}`)}`,
+          response,
+          detail: {
+            candidateErrors: candidateErrors,
+          },
+        });
+      }
+    }
+
+    // Pick the first valid candidate.
+    let selected: Candidate<any> | undefined;
+    for (const candidate of response.candidates) {
+      if (isValidCandidate(candidate, tools || [])) {
+        selected = candidate;
+        break;
+      }
+    }
+
+    if (!selected) {
+      throw new Error('No valid candidates found');
+    }
+
+    const toolCalls = selected.message.content.filter(
+      (part) => !!part.toolRequest
+    );
+    if (input.returnToolRequests || toolCalls.length === 0) {
+      return response.toJSON();
+    }
+    const toolResponses: ToolResponsePart[] = await Promise.all(
+      toolCalls.map(async (part) => {
+        if (!part.toolRequest) {
+          throw Error(
+            'Tool request expected but not provided in tool request part'
+          );
+        }
+        const tool = tools?.find(
+          (tool) => tool.__action.name === part.toolRequest?.name
+        );
+        if (!tool) {
+          throw Error('Tool not found');
+        }
+        return {
+          toolResponse: {
+            name: part.toolRequest.name,
+            ref: part.toolRequest.ref,
+            output: await tool(part.toolRequest?.input),
+          },
+        };
+      })
+    );
+    input.history = request.messages;
+    input.history.push(selected.message);
+    input.prompt = toolResponses;
+    return await generateAction(input);
```

Comment:
Can you copy instead of mutating here?

```suggestion
    const nextRequest = {
      ...input,
      history: [...request.messages, selected.message],
      prompt: toolResponses,
    }
    return await generateAction(nextRequest);
```